src/output-json-verdict.c

```diff
@@ -0,0 +1,228 @@
+ *
+ */
+
+#include "action-globals.h"
+
+#include "output.h"
+#include "output-json.h"
+#include "output-json-alert.h"
+#include "output-json-verdict.h"
+
+#include "packet.h"
+
+#include "suricata.h"
+#include "suricata-common.h"
+
+#define MODULE_NAME "JsonVerdictLog"
+
+typedef struct JsonVerdictOutputCtx_ {
+    uint8_t flags;
+    OutputJsonCtx *eve_ctx;
+} JsonVerdictOutputCtx;
+
+typedef struct JsonVerdictLogThread_ {
+    JsonVerdictOutputCtx *verdict_ctx;
+    OutputJsonThreadCtx *ctx;
+} JsonVerdictLogThread;
+
+void GetVerdictJsonInfo(JsonBuilder *jb, const Packet *p)
+{
+    jb_open_object(jb, "verdict");
+    /* add verdict info */
+    if (PacketCheckAction(p, ACTION_DROP) && EngineModeIsIPS()) {
+        JB_SET_STRING(jb, "action", "drop");
+    } else if (PacketCheckAction(p, ACTION_REJECT_ANY)) {
+        // check rule to define type of reject packet sent
+        if (EngineModeIsIPS()) {
+            JB_SET_STRING(jb, "action", "drop");
+        } else {
+            JB_SET_STRING(jb, "action", "alert");
+        }
+        if (p->action & ACTION_REJECT || p->action) {
+            JB_SET_STRING(jb, "reject-target", "source");
+        } else if (p->action & ACTION_REJECT_DST) {
+            JB_SET_STRING(jb, "reject-target", "destination");
+        } else if (p->action & ACTION_REJECT_BOTH) {
+            JB_SET_STRING(jb, "reject-target", "both");
+        }
+        JB_SET_STRING(jb, "reject", "[tcp-reset, icmp-prohib, user defined]");
+    } else if (PacketCheckAction(p, ACTION_PASS)) {
+        JB_SET_STRING(jb, "action", "pass");
+    } else {
+        // TODO make sure we don't have a situation where this wouldn't work
+        JB_SET_STRING(jb, "action", "alert");
+    }
+
+    /* Close verdict */
+    jb_close(jb);
+
+    return;
+}
+
+static int VerdictJson(JsonVerdictLogThread *vlt, const Packet *p)
+{
+    JsonVerdictOutputCtx *verdict_ctx = vlt->verdict_ctx;
+
+    JsonAddrInfo addr = json_addr_info_zero;
+    JsonAddrInfoInit(p, LOG_DIR_PACKET, &addr);
+
+    JsonBuilder *jb = CreateEveHeader(p, LOG_DIR_PACKET, "verdict", &addr, verdict_ctx->eve_ctx);
+
+    if (unlikely(jb == NULL))
+        return TM_ECODE_OK;
+
+    GetVerdictJsonInfo(jb, p);
+
+    OutputJsonBuilderBuffer(jb, vlt->ctx);
+    jb_free(jb);
+
+    return TM_ECODE_OK;
+}
+
+/**
+ * \brief   Log the final verdict for a packet, based on matched rules
+ *
+ * \param tv           Pointer to the current thread variables
+ * \param thread_data  Pointer to the verdict log structure
+ * \param p            Pointer to the packet that stores the info on verdict
+ *
+ * \return 0 on success
+ */
+static int JsonVerdictLogger(ThreadVars *tv, void *thread_data, const Packet *p)
+{
+    JsonVerdictLogThread *vlt = thread_data;
+
+    int r = VerdictJson(vlt, p);
+
+    if (r < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int JsonVerdictLogCondition(ThreadVars *tv, void *thread_data, const Packet *p)
+{
+    if (p->alerts.cnt > 0) {
```

Comment:
so verdict records are only written for alerts... this excludes drops from outside the detect engine (e.g. exception policy), also "pass" would be excluded, right?

Reply:
Right on cue for both aspects, I hadn't considered them >__<'